ixdiagnose/plugins/base.py

```diff
@@ -0,0 +1,91 @@
+import os
+import time
+import traceback
+
+from typing import List
+
+from ixdiagnose.utils.formatter import dumps
+from ixdiagnose.utils.middleware import get_middleware_client
+from ixdiagnose.utils.paths import get_plugin_base_dir
+
+from .metrics import Metric
+
+
+class Plugin:
+
+    metrics: List[Metric] = []
+    raw_metrics: List[Metric] = []
+    serializable_metrics: List[Metric] = []
+    name: str = NotImplementedError
+
+    def __init__(self):
+        self.debug_report: dict = {}
+
+        assert type(self.name) is str and bool(self.name) is True
+        for name in ('metrics', 'raw_metrics', 'serializable_metrics'):
+            metric_val = getattr(self, name)
+            assert type(metric_val) is list
+            assert all(isinstance(metric, Metric) for metric in metric_val)
+
+    @property
+    def output_dir(self) -> str:
+        return os.path.join(get_plugin_base_dir(), self.name)
+
+    def metrics_to_execute(self):
+        return self.metrics + self.raw_metrics
```

Comment:
Is this supposed to be hardcoded ?

Reply:
It is not..i'll be adding support for configuration extending the configuration class we have and the same will be exposed over CLI